### PR TITLE
feat!: use NodeJS 18 runtime

### PR DESCRIPTION
**Pull request by sobolk**
#### Description of changes

This PR:
- changes NodeJS runtime to 18 in packaged binary
- uses NodeJS18 image for circle ci linux jobs
- removes .NET Core 3.1 support (it reached EOL in December 2022)

Companion PR https://github.com/aws-amplify/amplify-cli/pull/12203

#### Description of how you validated changes

https://app.circleci.com/pipelines/github/aws-amplify/amplify-cli?branch=run-e2e%2Fsobkamil%2Fbump-pkg-node

Manual smoke tests on AL2.

https://app.circleci.com/pipelines/github/aws-amplify/amplify-cli?branch=run-e2e%2Fsobkamil%2Fbump-pkg-node-x%2Frun-all-tests

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies
- [ ] [Pull request labels](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#labels) are added

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

BREAKING CHANGE: Updated NodeJS runtime to 18.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -151,6 +151,7 @@
     "**/jake/async": "^2.6.4",
     "**/json5": "^2.2.3",
     "**/nth-check": "^2.0.1",
+    "**/pkg/prebuild-install": "7.1.1",
```
Comment (sobolk) [inline package.json:154]: Fixes. 
image

We might need to keep it until https://github.com/vercel/pkg/commit/e3ac4902e9cb023f0ac2a596e21ff9166d2e268c ships.

File: packages/amplify-cli/bin/amplify [hunk 2]
```diff
@@ -4,6 +4,8 @@
 // this file should be a duplicate of amplify.production.template except for the CLI_ENV value
 process.env.CLI_ENV = 'development';
 
+process['noDeprecation'] = true;
```
Comment (sobolk) [inline packages/amplify-cli/bin/amplify:7]: image

Comment (sobolk) [reply packages/amplify-cli/bin/amplify:7]: See https://nodejs.org/api/process.html#process_process_nodeprecation

File: .circleci/config.base.yml [hunk 3]
```diff
@@ -1354,3 +1356,17 @@ commands:
...
+      - run:
+          command: wget https://nodejs.org/download/release/v18.15.0/node-v18.15.0-x64.msi -P C:\Users\circleci\Downloads\
+          shell: cmd.exe
+      - run: MsiExec.exe /i C:\Users\circleci\Downloads\node-v18.15.0-x64.msi /qn
+      - run:
+          command: |
+            Start-Process powershell -verb runAs -Args "-start GeneralProfile"
+            nvm install 18.15.0
+            nvm use 18.15.0
+      - run: node --version
```
Comment (sobolk) [inline .circleci/config.base.yml:1372]: https://support.circleci.com/hc/en-us/articles/8325075309339-How-to-install-NPM-on-Windows-images

File: .circleci/config.base.yml [hunk 4]
```diff
@@ -16,7 +16,7 @@ parameters:
 executors:
   w_medium: &windows-e2e-executor-medium
     machine:
-      image: 'windows-server-2019-vs2019:stable'
+      image: 'windows-server-2022-gui:current'
```
Comment (sobolk) [inline .circleci/config.base.yml:19]: I was experimenting with installing Node 18 on old image too.
However I was running into issues with older version of nvm and conflicts with pre-installed node 14.
New image comes with fixed version of nvm but without node nor npm nor yarn, but at least installing them works as expected.

File: package.json [hunk 5]
```diff
@@ -46,7 +46,7 @@
     "setup-dev": "yarn dev-build && yarn rm-dev-link && yarn link-dev && yarn rm-aa-dev-link && yarn link-aa-dev",
     "split-e2e-tests": "yarn ts-node ./scripts/split-e2e-tests.ts && git add .circleci/config.yml",
     "test-changed": "nx affected --target=test",
-    "test-ci": "nx run-many --target=test --all -- --ci --runInBand",
+    "test-ci": "nx run-many --target=test --all -- --ci --runInBand --parallel=4",
```
Comment (sobolk) [inline package.json:49]: Had to ease resource pressure here as it became flaky. Other steps seem to be fine with default 8.

File: packages/amplify-migration-tests/src/__tests__/migration_tests_v10/dotnet_runtime_update_migration.test.ts [hunk 6]
```diff
@@ -1,88 +0,0 @@
...
-  functionMockAssert,
-  generateRandomShortId,
-} from '@aws-amplify/amplify-e2e-core';
-import { versionCheck, allowedVersionsToMigrateFrom } from '../../migration-helpers';
-import { initJSProjectWithProfileV10 } from '../../migration-helpers-v10/init';
-import { AmplifyCategories, JSONUtilities, pathManager } from 'amplify-cli-core';
-
-/**
- * These tests check that functions created with .NET Core 3.1 templates keep working after we drop support to create them.
- */
```
Comment (sobolk) [inline packages/amplify-migration-tests/src/__tests__/migration_tests_v10/dotnet_runtime_update_migration.test.ts:19]: We can't have .NET Core 3.1 in cci image that has node 18. 
We no longer provision .NET Core 3.1. too as we switched to .NET 6.
Therefore this sacrifice...

File: packages/amplify-console-integration-tests/src/consoleHosting/utils.ts [hunk 7]
```diff
@@ -82,7 +82,7 @@ export async function createTestProject(): Promise<string> {
   const projectName = path.basename(projRoot);
   const projectDir = path.dirname(projRoot);
 
-  spawnSync('npx', ['create-react-app', '--scripts-version', '4.0.3', projectName], { cwd: projectDir });
+  spawnSync('npx', ['create-react-app', '--scripts-version', '5.0.1', projectName], { cwd: projectDir });
```
Comment (sobolk) [inline packages/amplify-console-integration-tests/src/consoleHosting/utils.ts:85]: This update is related to https://stackoverflow.com/questions/69692842/error-message-error0308010cdigital-envelope-routinesunsupported .
I.e. react scripts 4.x are outdated and bring webpack that relies on older ssl.

File: package.json [hunk 8]
```diff
@@ -46,7 +46,7 @@
     "setup-dev": "yarn dev-build && yarn rm-dev-link && yarn link-dev && yarn rm-aa-dev-link && yarn link-aa-dev",
     "split-e2e-tests": "yarn ts-node ./scripts/split-e2e-tests.ts && git add .circleci/config.yml",
     "test-changed": "nx affected --target=test",
-    "test-ci": "nx run-many --target=test --all -- --ci --runInBand",
+    "test-ci": "nx run-many --parallel=4 --target=test --all -- --ci --runInBand",
```
Comment (awsluja) [inline package.json:49]: why run parallel in NX but runInBand at the jest layer?

**Review by sdstolworthy — APPROVED** [review]

**Review by awsluja — APPROVED** [review]

**Review by lazpavel — APPROVED** [review]

File: package.json (hunk 8 above)
Comment (sobolk) [reply package.json:49]: That comes from this PR https://github.com/aws-amplify/amplify-cli/pull/11917 .
I.e. using both nx and jest parallelization led to spin-locks.
In this PR we decrease default 8 parallelization to 4 for that single job due to slightly higher memory pressure (the alternative was to bump instance size)

Comment (sobolk) [reply package.json:49]: the default is here for reference https://github.com/aws-amplify/amplify-cli/blob/78590b8d4face7bbc0b174337b77b118affc525b/nx.json#L9

**Pull request merged**